src/meta/meta_table.c

```diff
@@ -64,44 +64,79 @@ __wt_metadata_open(WT_SESSION_IMPL *session)
 	if (F_ISSET(btree, WT_BTREE_NO_LOGGING))
 		F_CLR(btree, WT_BTREE_NO_LOGGING);
 
-	/* The metadata handle doesn't need to stay locked -- release it. */
-	return (__wt_session_release_btree(session));
+	return (0);
 }
 
 /*
  * __wt_metadata_cursor --
- *	Opens a cursor on the metadata.
+ *	Returns the session's cached metadata cursor, unless it's in use, in
+ * which case it opens and returns another metadata cursor.
  */
 int
-__wt_metadata_cursor(
-    WT_SESSION_IMPL *session, const char *config, WT_CURSOR **cursorp)
+__wt_metadata_cursor(WT_SESSION_IMPL *session, WT_CURSOR **cursorp)
 {
-	WT_DATA_HANDLE *saved_dhandle;
-	WT_DECL_RET;
-	bool is_dead;
-	const char *cfg[] =
-	    { WT_CONFIG_BASE(session, WT_SESSION_open_cursor), config, NULL };
+	WT_CURSOR *cursor;
 
-	saved_dhandle = session->dhandle;
-	WT_ERR(__wt_metadata_open(session));
+	/*
+	 * If we don't have a cached metadata cursor, or it's already in use,
+	 * we'll need to open a new one.
+	 */
+	cursor = NULL;
+	if (session->meta_cursor == NULL ||
+	    F_ISSET(session->meta_cursor, WT_CURSTD_META_INUSE)) {
+		WT_RET(__wt_metadata_cursor_open(session, NULL, &cursor));
+		if (session->meta_cursor == NULL) {
+			session->meta_cursor = cursor;
+			cursor = NULL;
+		}
+	}
 
-	session->dhandle = session->meta_dhandle;
+	/*
+	 * If there's no cursor return, we're done, our caller should have just
+	 * been triggering the creation of the session's cached cursor. There
+	 * should not be an open local cursor in that case, but caution doesn't
+	 * cost anything.
+	 */
+	if (cursorp == NULL)
+		return (cursor == NULL ? 0 : cursor->close(cursor));
 
-	/* 
-	 * We use the metadata a lot, so we have a handle cached; lock it and
-	 * increment the in-use counter once the cursor is open.
+	/*
+	 * If the cached cursor is in use, return the newly opened cursor, else
+	 * mark the cached cursor in use and return it.
 	 */
-	WT_ERR(__wt_session_lock_dhandle(session, 0, &is_dead));
+	if (F_ISSET(session->meta_cursor, WT_CURSTD_META_INUSE))
+		*cursorp = cursor;
+	else {
+		*cursorp = session->meta_cursor;
+		F_SET(session->meta_cursor, WT_CURSTD_META_INUSE);
+	}
+	return (0);
+}
 
-	/* The metadata should never be closed. */
-	WT_ASSERT(session, !is_dead);
+/*
+ * __wt_metadata_cursor_release --
+ *	Release a metadata cursor.
+ */
+int
+__wt_metadata_cursor_release(WT_SESSION_IMPL *session, WT_CURSOR **cursorp)
+{
+	WT_CURSOR *cursor;
 
-	WT_ERR(__wt_curfile_create(session, NULL, cfg, false, false, cursorp));
-	__wt_cursor_dhandle_incr_use(session);
+	WT_UNUSED(session);
 
-	/* Restore the caller's btree. */
-err:	session->dhandle = saved_dhandle;
-	return (ret);
+	if ((cursor = *cursorp) == NULL)
+		return (0);
+	*cursorp = NULL;
+
+	/*
+	 * If using the session's cached metadata cursor, clear the in-use flag
+	 * and reset it, otherwise, discard the cursor.
+	 */
+	if (F_ISSET(cursor, WT_CURSTD_META_INUSE)) {
```

Comment:
I'd feel more comfortable comparing the cursor pointers than the flag here, or at least asserting that they are the same if we expect them to be. i.e: This code assumes that acquisition/release are entirely nested.